### PR TITLE
dmd.dmodule: fix memory leak in lookForSourceFile

### PR DESCRIPTION
**Pull request by andy-k**
`lookForSourceFile` calls `forceExt`, but did not always call `free` as [`forceExt` suggests](https://github.com/dlang/dmd/blob/e09e97e0b8468bc1ce238bcb8de95f80a5beb430/src/dmd/root/filename.d#L551-L565).

### CODE REVIEW

File: src/dmd/backend/cdef.d [hunk 1]
```diff
@@ -258,6 +258,7 @@ public import dmd.root.longdouble : targ_ldouble = longdouble;
 
 // Extract most significant register from constant
 int REGSIZE();
+extern (C)
```
Comment (jacob-carlborg) [inline src/dmd/backend/cdef.d]: Why this change?

Comment (andy-k) [reply src/dmd/backend/cdef.d]: On OSX I wasn't able to link the code as cloned:

```
$ make -f posix.mak test
...
Undefined symbols for architecture x86_64:
  "_MSREG", referenced from:
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: linker exited with status 1
make[1]: *** [../generated/osx/release/64/dmd] Error 1
```

This change, in a separate commit, seems to fix it.

Would it be better if I split them into two pull requests? This would leave my remaining changes for this pull request not locally tested.

Comment (jacob-carlborg) [reply src/dmd/backend/cdef.d]: It works fine to build for me locally. Please try to clean and do a full rebuild. I'm using DMD 2.085.0 as the host compiler.

Comment (andy-k) [reply src/dmd/backend/cdef.d]: Thanks, it works now after clean. Removing this commit from this pull request.

File: src/dmd/dmodule.d [hunk 2]
```diff
@@ -66,9 +66,11 @@ private const(char)[] lookForSourceFile(const(char)[] filename)
     const sdi = FileName.forceExt(filename, global.hdr_ext.toDString());
     if (FileName.exists(sdi) == 1)
         return sdi;
+    scope(exit) FileName.free(sdi.ptr);
```
Comment (WalterBright) [inline src/dmd/dmodule.d]: the `scope(exit)` is not necessary here

File: src/dmd/dmodule.d [hunk 3]
```diff
@@ -66,9 +66,11 @@ private const(char)[] lookForSourceFile(const(char)[] filename)
...
     const sd = FileName.forceExt(filename, global.mars_ext.toDString());
     if (FileName.exists(sd) == 1)
         return sd;
+    scope(exit) FileName.free(sd.ptr);
```
Comment (WalterBright) [inline src/dmd/dmodule.d]: or here

File: src/dmd/dmodule.d [hunk 4]
```diff
@@ -506,7 +508,10 @@ extern (C++) final class Module : Package
         /* Look for the source file
          */
         if (const result = lookForSourceFile(filename))
+        {
+            scope(exit) FileName.free(result.ptr);
```
Comment (WalterBright) [inline src/dmd/dmodule.d]: remove `scope(exit)` and move the rest to after line 513. DMD doesn't use exceptions, so this wouldn't be a problem.

File: src/dmd/dmodule.d (hunk 2 above)
Comment (andy-k) [reply src/dmd/dmodule.d]: Both of these `scope(exit)` are necessary, because `sdi` and `sd` may be used [in the `for`-loop below](https://github.com/dlang/dmd/blob/a032e3cfeb53add32fbb60e6652b1d3515d88e40/src/dmd/dmodule.d#L96-L101).

**Review by thewilsonator — APPROVED** [review]
